### PR TITLE
Added event_to_attr in register_events

### LINKED ISSUE #501: event_to_attr should be extensible for register_event
There should be a way to extend `event_to_attr` when registering new events.

### PR DESCRIPTION
**Pull request by anmolsjoshi**
Fixes #501  

Description:
* Added parameter in register_events to update state with Custom events
* Update State and Engine to initialize values using State.event_to_attr (removed hard coded epoch=0, iteration=0 in initialization of State)

Check list:
* [x] New tests are added (if a new feature is added)
* [x] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/engine/engine.py [hunk 1]
```diff
@@ -89,18 +91,19 @@ def __init__(self, process_function):
...
         This opens the door to make the :meth:`~ignite.engine.Engine.run` loop even more
         configurable.
 
-        By default, the events from :class:`~ignite.engine.Events` are registerd.
+        By default, the events from :class:`~ignite.engine.Events` are registered.
 
         Args:
             *event_names: An object (ideally a string or int) to define the
                 name of the event being supported.
+            event_to_attr (dict/object): A dictionary or object to map an event to a state attribute.
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: > dict/object): A dictionary or object to map an event to a state attribute.

This is not clear what is the type of event_to_attr.

File: ignite/engine/engine.py [hunk 2]
```diff
@@ -114,10 +117,15 @@ class Custom_Events(Enum):
 
             engine = Engine(process_function)
             engine.register_events(*Custom_Events)
```
Comment (vfdev-5) [inline ignite/engine/engine.py]: Maybe we can provide another example here with `event_to_attr`

File: ignite/engine/engine.py (hunk 1 above)
Comment (anmolsjoshi) [reply ignite/engine/engine.py]: Updated, event_to_attr should only be a dictionary, code includes check. 

Python 2.7 does not allow for regular parameter after * parameter, so had to update the code accordingly.

File: ignite/engine/engine.py (hunk 2 above)
Comment (anmolsjoshi) [reply ignite/engine/engine.py]: Provide example in docstring.

**Pull request merged**